lib/src/lifecycle_module.dart

```diff
@@ -274,36 +369,54 @@ abstract class LifecycleModule extends Object with Disposable {
 
   /// Public method to trigger the Module unload cycle.
   ///
-  /// Calls shouldUnload(), and, if that completes successfully,
-  /// continues to call onUnload() on the module and all registered child modules.
-  /// If unloading is rejected, this method will complete with an error.
+  /// Calls shouldUnload(), and, if that completes successfully, continues to
+  /// call onUnload() on the module and all registered child modules. If
+  /// unloading is rejected, this method will complete with an error.
+  ///
+  /// Initiates the unload process when the module is in the loaded or suspended
+  /// state. If the module is in the unloading or unloaded state a warning is
+  /// logged and the method is a noop. If the module is in any other state, a
+  /// StateError is thrown.
   Future<Null> unload() {
+    if (isUnloaded || isUnloading) {
+      return _buildNoopResponse(
+          isTransitioning: isUnloading,
+          methodName: 'unload',
+          currentState: LifecycleState.unloaded);
+    }
+
+    if (!(isLoaded || isSuspended)) {
+      return _buildIllegalTransitionResponse(
+          targetState: LifecycleState.unloaded,
+          allowedStates: [LifecycleState.loaded, LifecycleState.suspended]);
+    }
+
+    ShouldUnloadResult canUnload = shouldUnload();
```

Comment:
Can we call this something like `shouldUnloadResult` or something? `canUnload.shouldUnload` is bugging me...